### ISSUE TITLE
Outdated diagnostics on InsertLeave: refresh is triggered when completion item is selected

### ISSUE THREAD

**Issue by HansPinckaers** [neoclide/coc.nvim#1489]
## Result from CocInfo
```

 ## versions                                                            
                                                                        
 vim version: NVIM v0.5.0-330-g71ee46acc                                
 node version: v12.14.1                                                 
 coc.nvim version: 0.0.74                                               
 term: screen-256color                                                  
 platform: linux                                                        
                                                                        
 ## Messages                                                            
                                                                        
 ## Output channel: Pyright                                             
                                                                        
 Pyright language server starting                                       
 Starting service instance "MIL"                                        
 Fetching settings for workspace(s)                                     
 Received updated settings                                              
 No configuration file found.                                           
 Setting pythonPath for service "MIL": "/home/hans/miniconda3/bin/python"
 Searching for source files                                             
 Found 32 source files                                                  
                                                                        
 ## Output channel: Python                                              
                                                                        
 Starting Jedi Python language engine.                                  
                                                                        
                                                                        
                
```                                                        
                                                                        
## Describe the bug

As soon as I leave insert-mode and after competing certain items, a diagnostic hover text shows up with a outdated message (basically an error relevant to how the text looked before completion. For example:

```
self.<completion items>
self.settings() (accept completion)
<InsertLeave>
self.settings()  -> virtual test shows: Expected member name after '.'
<after roughtly half a second>
self.settings()
```
![cropped](https://user-images.githubusercontent.com/70747/73064989-baedeb80-3ea2-11ea-9fd3-44a0813bbec1.png)

## Reproduce the bug

For me this also happens in :CocConfig. If I'm accepting a completion item in insert mode.

You will need:

```
  "diagnostic.refreshOnInsertMode": false,
  "diagnostic.virtualText": true,
```

**Comment by oblitum** [comment]
You're running two language servers at the same time?

**Comment by HansPinckaers** [comment]
Yes, coc-python and coc-pyright, for the last I disabled the completions (disableLanguageServices).

**Comment by douglarek** [comment]
Yeah, I also encountered this issue, And it is normal when I use `vim` instead of `neovim`

**Comment by oblitum** [comment]
That's really confusing b/c OP is referring to a NeoVim-only feature (virtualtext), afaik.

**Comment by douglarek** [comment]
> That's really confusing b/c OP is referring to a NeoVim-only feature (virtualtext), afaik.

I did not mean `vim` supports **virtualtext**, I just say outdated diagnostics does not occur in `vim` but `neovim`

**Comment by oblitum** [comment]
Well, OP say it's needed.

**Comment by HansPinckaers** [comment]
Tbh I didn't test it without.

It looks like an InsertLeave event is triggered on selection of a completion item and then InsertEnter is triggered when you start typing again. 

So I guess this bug should also happen with other linter configurations.

**Comment by chemzqm** [comment]
Can't reproduce, could be old version of coc.nvim.

**Comment by HeitorBoschirolli** [comment]
It happens to me too. About reproducing the bug, for me, it starts to happen only after using neovim for a while; restarting it causes the bug to disappear for a while.

versions I'm using:
vim version: NVIM v0.4.2
node version: v10.16.3
coc.nvim version: 0.0.74-1ff50ec3b0

**Comment by sQVe** [comment]
I think this issue might be related to neoclide/coc-tsserver#116 in some way.

**Comment by zArubaru** [comment]
Seeing a similar issue (Vim) with tsserver and eslint. Still trying to diagnose when and how specifically it happens, as for now it seems a bit random.

**Comment by chemzqm** [comment]
Duplicate of #1385

**Comment by HansPinckaers** [comment]
Still having this issue, everything is up to date (using master). I'm not sure if it is related to #1385 as the 'old' errors do not persist. Also, I've only seen this happening with very fast linters.

This patch fixes it, but it is *extremely* hacky. However, it does make me believe there is some issue with ordering of events (like the onDidDiagnosticsClear event triggers a refreshBuffer but the linter has pushed errors based on outdated content?).

For me it is very easy to reproduce. If I type in :CocConfig and accept a completion via tab, then leave insert mode, I see a diagnostic-error flash like described above.

```typescript
diff --git src/diagnostic/manager.ts src/diagnostic/manager.ts
index f0ce02f..e95af2c 100644
--- src/diagnostic/manager.ts
+++ src/diagnostic/manager.ts
@@ -52,6 +52,7 @@ export class DiagnosticManager implements Disposable {
   public readonly buffers: DiagnosticBuffer[] = []
   private lastMessage = ''
   private floatFactory: FloatFactory
+  private leaveCounter = 0
   private collections: DiagnosticCollection[] = []
   private disposables: Disposable[] = []
   private timer: NodeJS.Timer
@@ -75,8 +76,8 @@ export class DiagnosticManager implements Disposable {
     events.on('InsertEnter', async () => {
       if (this.timer) clearTimeout(this.timer)
       this.floatFactory.close()
+      this.leaveCounter = 0
     }, null, this.disposables)
-
     events.on('InsertLeave', async bufnr => {
       this.floatFactory.close()
       let doc = workspace.getDocument(bufnr)
@@ -87,9 +88,12 @@ export class DiagnosticManager implements Disposable {
           doc.forceSync()
           await wait(50)
         }
-        let d = 300 - (Date.now() - this.lastChanageTs)
-        if (d > 0) await wait(d)
-        this.refreshBuffer(doc.uri)
+        this.leaveCounter = this.leaveCounter + 1
+        if (this.leaveCounter == 1) {
+          let d = 300 - (Date.now() - this.lastChanageTs)
+          if (d > 0) await wait(d)
+          this.refreshBuffer(doc.uri)
+        }
       }
     }, null, this.disposables)
 
@@ -620,6 +624,10 @@ export class DiagnosticManager implements Disposable {
   private refreshBuffer(uri: string): boolean {
     let { insertMode } = workspace
     if (insertMode && !this.config.refreshOnInsertMode) return
+
+    this.leaveCounter = this.leaveCounter + 1
+    if (this.leaveCounter < 3) return
+
     let buf = this.buffers.find(buf => buf.uri == uri)
     if (!buf) return
     let { displayByAle } = this.config
```
